hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OmTransportFactory.java

```diff
@@ -0,0 +1,63 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.om.protocolPB;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
+import org.apache.hadoop.hdds.conf.ConfigurationSource;
+import org.apache.hadoop.security.UserGroupInformation;
+
+/**
+ * Factory pattern to create object for RPC communication with OM.
+ */
+public interface OmTransportFactory {
+
+  OmTransport createOmTransport(ConfigurationSource source,
+      UserGroupInformation ugi, String omServiceId) throws IOException;
+
+  static OmTransport create(ConfigurationSource conf,
+      UserGroupInformation ugi, String omServiceId) throws IOException {
+    OmTransportFactory factory = createFactory();
+
+    return factory.createOmTransport(conf, ugi, omServiceId);
+  }
+
+  static OmTransportFactory createFactory() throws IOException {
+    ServiceLoader<OmTransportFactory> transportFactoryServiceLoader =
+        ServiceLoader.load(OmTransportFactory.class);
+    Iterator<OmTransportFactory> iterator =
```

Comment:
Good question. This is because we need a default and an option to choose a **different** implementation. 

If we create a service file for the default transport factory (`Hadoop3OmTransportFactory`) than it would be hard to override it. In case of the Hadoop2 client we would have two different implementation defined without an easy way to choose between them.

With this approach the default is always defined by line 50:61 but you have the option to define your own version.

Reply:
Question: what is alternative way other then having meta-inf/services to load class for OmTransportFactory?

Reply:
Currently there is no alternative way. I couldn't see any usage for any other approach, but if you see any use case, I am open to implement it. Let's create a new issue if you see some interesting direction...

Reply:
Then I am confused, what is the purpose of the code?
Because I think meta-inf/services is packaged along with jar right?